### PR TITLE
fix(eks): missing support for "InstanceTypes" attribute assignment for AL2023 AMIs

### PR DESCRIPTION
**Pull request by guessi**
### Issue # (if applicable)

Closes https://github.com/aws/aws-cdk/issues/29546

### Reason for this change

After #29335, `@aws-eks` should receive AL2023 support, despites the [GPU types are not yet supported](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html#cfn-eks-nodegroup-amitype) it should at least allow user to customize instance type. However, missing support for `NodegroupAmiType[]` causing validation error emit, so that user can only create [default instance types](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html#cfn-eks-nodegroup-instancetypes) ( `t3.medium` or `t4g.medium` ).

```
$ cat ./lib/cluster.ts
...
    cluster.addNodegroupCapacity("mng-al2023", {
      amiType: eks.NodegroupAmiType.AL2023_X86_64_STANDARD,
      instanceTypes: [new ec2.InstanceType("t3.medium")],
      ...
    });
...
```

```
$ npx cdk version
2.133.0 (build dcc1e75)
```

```
$  npx cdk synth

...
                                                                                                                                                                                ^
Error: The specified AMI does not match the instance types architecture, either specify one of AL2_X86_64, BOTTLEROCKET_X86_64, WINDOWS_CORE_2019_X86_64, WINDOWS_CORE_2022_X86_64, WINDOWS_FULL_2019_X86_64, WINDOWS_FULL_2022_X86_64 or don't specify any
    at new Nodegroup (.../node_modules/aws-cdk-lib/aws-eks/lib/managed-nodegroup.js:1:3921)
    at Cluster.addNodegroupCapacity (.../node_modules/aws-cdk-lib/aws-eks/lib/cluster.js:1:19807)
    at EksCluster.createManagedNodeGroups (.../lib/cluster.ts:85:13)
    at new EksCluster (.../lib/cluster.ts:30:10)
    at Object.<anonymous> (.../bin/eks-basic.ts:10:1)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module.m._compile (.../node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Object.require.extensions.<computed> [as .ts] (.../node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1207:32)

Subprocess exited with error 1
```

### Description of changes

Add `eks.NodegroupAmiType.AL2023_X86_64_STANDARD` and `eks.NodegroupAmiType.AL2023_ARM_64_STANDARD` support for node group module.

### Description of how you validated changes

```
$ npx jest aws-eks/test/cluster.test.ts
 PASS  aws-eks/test/cluster.test.ts (37.298 s)
...

=============================== Coverage summary ===============================
Statements   : 48.86% ( 10077/20621 )
Branches     : 27.88% ( 2388/8565 )
Functions    : 33.03% ( 1509/4568 )
Lines        : 49.71% ( 9905/19923 )
================================================================================
Jest: "global" coverage threshold for statements (55%) not met: 48.86%
Jest: "global" coverage threshold for branches (35%) not met: 27.88%
Test Suites: 1 passed, 1 total
Tests:       119 passed, 119 total
Snapshots:   0 total
Time:        41.064 s
```

```
$ npx jest aws-eks/test/nodegroup.test.ts
 RUNS  aws-eks/test/nodegroup.test.ts
...
=============================== Coverage summary ===============================
Statements   : 43.94% ( 9062/20621 )
Branches     : 22.08% ( 1892/8565 )
Functions    : 27.16% ( 1241/4568 )
Lines        : 44.8% ( 8927/19923 )
================================================================================
Jest: "global" coverage threshold for statements (55%) not met: 43.94%
Jest: "global" coverage threshold for branches (35%) not met: 22.08%
Test Suites: 1 passed, 1 total
Tests:       59 passed, 59 total
Snapshots:   0 total
Time:        23.334 s
```

### Checklist
- [X] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts:567]: It's the key to the issue, missing definitions for AmiType: `AL2023_[ARM_64|X86_64]_STANDARD`.

**Comment by pahud** [comment]
Thank you. This PR LGTM. Are you able to update the `integ.eks-cluster.ts`?

And you'll need to fix the CI before we can move this PR forward.

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (pahud) [inline packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:21]: If you use 1.29 here, you should provide  `kubectlLayer` for kubectl lambda layer 1.29 as well. I feel updating this file may cause some additional issues to fix and I tend to have another separate PR to do that.

If you look at my PR draft I actually created a new integ test just for AL2023 and leave the huge cluster.test.ts untouched for this PR. What do you think?

https://github.com/aws/aws-cdk/pull/29550/files#diff-fb4c143ff72c5e6ae4920d14fa54a0e15acd140583a42c768d839e46d2331ff9

**Comment by guessi** [comment]
Checking log...

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (pahud) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:21]: But if you can successfully deploy cluster.test.ts that's even better! All we need to do is to add `kubectlLayer` with 1.29 for the cluster creation.

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (pahud) [inline packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:30]: I would suggest leave this untouched. I know V2_4_1 is not the latest version but I would prefer bump this version when we really have to.

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (guessi) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:21]: thanks for pointing out why integ test would failed, fixing.

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (guessi) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:30]: sure thing, I will revert this change.

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
Comment (pahud) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:21]: Hi @guessi 

Let me elaborate a little bit.

If you look at the cluster.test.ts in the main branch
https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
The current cluster version is 1.25 with NO `kubectlLayer` specified, this should pass the unit testing but in real deployment, leaving `kubectlLayer` unspecified is not recommended. I am fine if we leave it unspecified in the unit tests but if you need to update [integ.eks-cluster.ts](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-cluster.ts) and verify it with 1.29, you will need to update [this](https://github.com/aws/aws-cdk/blob/c9fb4f739f3aacf669cdd38b8431695811be5e92/packages/%40aws-cdk-testing/framework-integ/test/aws-eks/test/integ-tests-kubernetes-version.ts#L6-L14) and it would be a long huge deployment. I guess you probably will fail the integ.eks-cluster.ts somewhere. But if you are really interested to dive deep, it would be great to have integ.eks-cluster.ts to support 1.29.

In my PR. I selected another path - leave integ.eks-cluster.ts untouched and create a new integ test to make sure AL2023 would deploy with the latest k8s 1.29. It's very opinionated but I am good with both path.

Comment (guessi) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:21]: Understood!

File: packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts:672]: To ensure MGN with LT could capture invalid instance type exception.

File: packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts:694]: To ensure MGN with LT could capture invalid instance type exception.

File: packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/test/nodegroup.test.ts:172]: To ensure MNG could be created with no instance type specified.

File: packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts:573]: It's the core to the issue, missing definition for AL2023.

File: packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts:568]: It's the core to the issue, missing definition for AL2023.

File: packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ-tests-kubernetes-version.ts
Comment (guessi) [inline packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ-tests-kubernetes-version.ts:10]: To make sure it could compatible with existing integ tests.

**Comment by guessi** [comment]
@pahud please have a look while you are free, thanks!

**Comment by pahud** [comment]
LGTM. We'll need another review and approval from the maintainers.

**Review by pahud — COMMENTED** [review]
LGTM. We'll need another review and approval from the maintainers.

**Review by GavinZZ — COMMENTED** [review]
Can we please add an example to [README.md](https://github.com/aws/aws-cdk/blob/8df2823037553d3f4c1ce28720a883c05b68ee85/packages/aws-cdk-lib/aws-eks/README.md#L898) with AL2023_ARM_64_STANDARD usage example?

File: packages/aws-cdk-lib/aws-eks/README.md
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/README.md:190]: Reason for why removing this line
* The section here should focus on `addNodegroupCapacity` but not `amiType`.
* Instance type here `m5.large` does not support `GPU` which is misleading.

**Comment by guessi** [comment]
@GavinZZ I've updated the `README.md` as separated commit, please review again and check for the comment above.

File: packages/aws-cdk-lib/aws-eks/README.md
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/README.md:225]: Discussion: should we address AL2023 here? or maybe keep it as default `AL2_ARM_64`.

File: packages/aws-cdk-lib/aws-eks/README.md
Comment (guessi) [inline packages/aws-cdk-lib/aws-eks/README.md:217]: Discussion: should we address AL2023 here? or maybe keep it as default `AL2_X86_64`.

**Comment by guessi** [comment]
Umm... weird, let me trigger PR check again

```
aws-cdk-lib.aws_eks-README-L913.ts:46:3 - error TS2353: Object literal may only specify known properties, and 'amiType' does not exist in type 'AutoScalingGroupCapacityOptions'.

46   amiType: eks.NodegroupAmiType.AL2023_ARM_64_STANDARD,
```

**Comment by guessi** [comment]
@GavinZZ I've updated the `README.md` as separated commit, please review once again, thanks!

As discussed with @pahud offline, I'm here leaving change for [ARM64 Support](https://github.com/aws/aws-cdk/blob/v2.133.0/packages/aws-cdk-lib/aws-eks/README.md#arm64-support) section untouched as making change to `addAutoScalingGroupCapacity()` would need to make change to the enum of [MachineImageType](https://github.com/aws/aws-cdk/blob/v2.133.0/packages/aws-cdk-lib/aws-eks/lib/cluster.ts#L2459-L2471) and the would trigger integration tests for `cluster.ts`, which is not we are going to see at this PR.

I can see there are many `if... else...` blocks for `MachineImageType` detection logic for bootstraping

* https://github.com/aws/aws-cdk/blob/v2.133.0/packages/aws-cdk-lib/aws-eks/lib/cluster.ts#L1152-L1157
* https://github.com/aws/aws-cdk/blob/v2.133.0/packages/aws-cdk-lib/aws-eks/lib/cluster.ts#L1762-L1808

I think `AL2023` support for `addAutoScalingGroupCapacity()` would better to be done in separated PR, at this PR here we should focus on MNG first... WDYT?

**Review by GavinZZ — DISMISSED** [review]

**Comment by GavinZZ** [comment]
@guessi can you please fix the merge conflict and this PR should be good to go.

**Comment by guessi** [comment]
@GavinZZ Just rebased, should have no conflict now 👍

**Review by GavinZZ — APPROVED** [review]

**Pull request merged**
